### ISSUE TITLE
Type error involving list/tuple/enum type.

### ISSUE THREAD

**Issue by danielpopescu** [microsoft/pyright#7090]
Pyright: 1.1.348. foo1 function is fine, but foo reports a type error.... it seems incorrect... functions do work correctly. ty. 

from __future__ import annotations

from enum import IntEnum
from typing import Tuple

class Error(IntEnum):
    BAD = 0
    GOOD = 1

Elem = Tuple[int, Error]

def foo(k: int) -> list[Elem]:
    return k * [(0, Error.GOOD)]

def foo1(k: int) -> list[float]:
    return k * [1.0]

**Comment by erictraut** [comment]
When posting code samples in github, you can surround the code with a triple back tick so it formats correctly. Here's your properly-formatted code sample for reference.

```python
from __future__ import annotations

from enum import IntEnum
from typing import Tuple

class Error(IntEnum):
    BAD = 0
    GOOD = 1

Elem = Tuple[int, Error]

def foo(k: int) -> list[Elem]:
    return k * [(0, Error.GOOD)]

def foo1(k: int) -> list[float]:
    return k * [1.0]
```

**Comment by erictraut** [comment]
Pyright is working as intended in this case, so I don't consider it a bug. This is a case where you need to provide additional type information for the type checker to properly evaluate the expression's type.

```python
def foo(k: int) -> list[Elem]:
    x: list[tuple[int, Error]] = [(0, Error.GOOD)]
    return k * x
```

The problem is that binary expressions in Python (such as `k * x`) are not simple magic method calls. They involve a complex series of calls to multiple magic methods (e.g. `__mul__` followed optionally by `__rmul__`). To handle this, a static type checker must evaluate both the types of the left-hand and right-hand operand types without the benefit of any additional context, such as the expected type of the final operation. In your code sample, the type of the right-hand operand (`[(0, Error.GOOD)]`) evaluates to `list[tuple[Literal[0], Literal[Error.GOOD]]`. Because a `list` is invariant in its type parameter, this is not compatible with `list[Elem]`.
